Adgs1408 driver

mirceacaprioru opened the pull request:
mux:adgs1408/1409: New driver for Analog Devices ADGS1408/1409 mux  …
This patch adds basic support for Analog Device ADGS1408/09 SPI mux
controller.

The device is probed and set to a disabled state. It uses the new mux
controller framework.

Signed-off-by: Mircea Caprioru <mircea.caprioru@analog.com>

lclausen-adi commented on arch/arm/boot/dts/bcm2710-rpi-3-b.dts:
```diff
@@ -15,6 +16,17 @@
 		serial0 = &uart1;
 		serial1 = &uart0;
 	};
+	adc-mux@3 {
```
The changes to the RPI dts should be in a separate commit. Since this is just an example of how to use the driver and not part of the driver itself.

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+
+static int adgs140x_spi_reg_write(struct spi_device *spi,
+				u8 reg_addr,
+				u8 reg_data)
```
Maybe put reg_addr and reg_data on the same line, there is enough space.

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+		return ret;
+
+	if (cells)
+		dev_info(dev, "2x  single pole/ single throw mux registered\n");
```
I'd change this to dev_dbg, or maybe even completely remove it. If every driver that is probed prints a message you'd have bootlog that keeps scrolling for days.

lclausen-adi commented on drivers/mux/Makefile:
```diff
@@ -4,10 +4,12 @@
 
 mux-core-objs			:= core.o
 mux-adg792a-objs		:= adg792a.o
+mux-adgs140x-objs               := adgs140x.o
```
Should use tabs instead of spaces for alignment.

commodo commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+	tx_buf[0] = reg_addr;
+	tx_buf[1] = reg_data;
+
+	ret = spi_write_then_read(spi, &tx_buf[0], n_tx, NULL, 0);
```
nitpick: using an extra variable ( `n_tx` ) seems a bit much ; I'd just add `2` or you could use `sizeof(tx_buf)`

also, in this case you can do directly  `return  spi_write_then_read(spi, &tx_buf[0], n_tx, NULL, 0);`
that way, you don't need the `ret` variable either

lclausen-adi replied:
Just ```return spi_write_then_read(...)``` no need for the extra variable.

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+			ret = adgs140x_spi_reg_write(spi, ADGS1408_SW_DATA,
+							ADGS1409_MUX(state));
+	}
+
```
This could be slightly simplified since you are writing to the same register on all the branches. Have a variable that will hold the register value and then just do one write at the end of the function.

commodo commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+	struct spi_device *spi = to_spi_device(mux->chip->dev.parent);
+	u8 ret;
+
+	if (mux->chip->controllers == 1) {
```
this `if` block seems a bit too verbose;
you could compress it a bit like

```
if (state == MUX_IDLE_DISCONNECT)
    data = ADGS1408_DISABLE;
else
    data = (mux->chip->controllers == 1) ? ADGS1408_MUX(state) : ADGS1409_MUX(state);

return adgs140x_spi_reg_write(spi, ADGS1408_SW_DATA, data);
```

lclausen-adi replied:
No ternary statements. Upstream does not like that. Just classic if (...) ... else if (...) ... else ...

commodo commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,158 @@
+
+	mux_chip->ops = &adgs140x_ops;
+
+	ret = adgs140x_spi_reg_write(spi, 0x01, ADGS1408_DISABLE);
```
use `ADGS1408_SW_DATA` here (instead of 0x01), since it's already defined

commodo commented:
there's a new err from checkpatch ; you need to format your email correctly
`Signed-off-by: Mircea Caprioru mircea.caprioru@analog.com` ==> `Signed-off-by: Mircea Caprioru <mircea.caprioru@analog.com>`

commodo commented on Documentation/devicetree/bindings/mux/adi,adgs140x.txt:
```diff
@@ -0,0 +1,39 @@
+Bindings for Analog Devices ADGS1408/1409 8:1/Dual 4:1 Mux
```
The comment is: 
`This should address the checkpatch.pl warning.`
Ideally this should focus on detailing the change, which is adding dt-bindings documentation

lclausen-adi commented:
> This should address the checkpatch.pl warning.

should not be part of the commit message.

The reason why your write the documentation is not to silence checkpatch. It is to provide documentation to others how to write bindings for the device. The commit message for the bindings give a short introduction to what kind of part this is and then point out any special things about the bindings. In this case you can say that the bindings follow the standard SPI and mux bindings and do not require any additional custom properties. This gives the reviewer and idea what to expect.

commodo commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,137 @@
+
+	if (state == MUX_IDLE_DISCONNECT)
+		reg = ADGS1408_DISABLE;
+	else
```
was there supposed to be a `ADGS1409_MUX(state)` else branch here ?

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,137 @@
+#define ADGS1408_SW_DATA       (0x01)
+#define ADGS1408_REG_READ(reg) (reg | 0x80)
+#define ADGS1408_DISABLE       (0x00)
+#define ADGS1408_MUX(state)    (((state & 0x07) << 1) + 1)
```
Maybe `| 1` instead of `+ 1`. Same result but makes more logical sense.

stefpopa commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,137 @@
+ *
+ * Licensed under the GPL-2.
+ */
+
```
You should use SPDX-License-Identifer following Documentation/process/license-rules.rst.
I  think this should be // SPDX-License-Identifier: GPL-2.0

mirceacaprioru commented:
Changelog V3 -> V4

- added new license comment
- removed state mask for SW_DATA configuration
- changed from +1 to  | 1 for better logical understanding

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,136 @@
+		}
+	}
+
+	ret = devm_mux_chip_register(dev, mux_chip);
```
Just return `devm_mux_chip_register(...)`.

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,136 @@
+#include <linux/property.h>
+
+#define ADGS1408_SW_DATA       (0x01)
+#define ADGS1408_REG_READ(reg) (reg | 0x80)
```
It's good practice to always add brackets around macro parameters when using them inside the macro to avoid accidental wrong expansion.

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,136 @@
+#define ADGS1408_SW_DATA       (0x01)
+#define ADGS1408_REG_READ(reg) (reg | 0x80)
+#define ADGS1408_DISABLE       (0x00)
+#define ADGS1408_MUX(state)    ((state << 1) | 1)
```
same here

lclausen-adi commented on drivers/mux/adgs140x.c:
```diff
@@ -0,0 +1,136 @@
+	tx_buf[0] = reg_addr;
+	tx_buf[1] = reg_data;
+
+	return spi_write_then_read(spi, &tx_buf[0], 2, NULL, 0);
```
just tx_buf and sizeof(tx_buf) instead of 2.
